app/models/grda_warehouse/client_file.rb

```diff
@@ -422,6 +424,24 @@ def sync_revokation_info(current_user)
       end
     end
 
+    def self.maintain_urls
+      where(url: nil).find_in_batches do |files|
```

Comment:
It seems like this logic is assuming the client_file can never change. Is that a safe assumption? We don't have any enforcement for that (such as a validation).

- option1: Could we just reprocess all files or would that be too expensive? I imagine we could calculate the url and skip the row if it matches. Also maybe it doesn't need to run every hour.
- option2: Add a rails lifecycle event to update the url if the attachment changes
- option3: Add a validation that explodes if the attachment is updated

Reply:
Good thoughts.  The warehouse files are immutable, but I don't think that's true in HMIS.  I'll update.  Probably best to just have it blank out the url field in a before_save hook.

This seemed light-weight enough to run every hour given that people may be adding files at any time and it should only process unprocessed files. I'm happy to move it to the every 4 hours schedule, but it should probably happen more than once a day to ensure we have as many as possible for each daily load to OP Analytics.